[#8864] Student: Feedback submission: Rank question: Improve explanation regarding ranking constraints

mg14777 opened the pull request:
Fixes #8864 

**Outline of Solution**
The presentation of information has been changed to follow Distribute points questions (PR [8836](https://github.com/TEAMMATES/teammates/pull/8836)). The new view looks like this:

![image](https://user-images.githubusercontent.com/11634699/40889768-6063c516-6764-11e8-9a23-1b6243385c23.png)

![image](https://user-images.githubusercontent.com/11634699/40889781-7c742cf0-6764-11e8-97dd-70acd3fef310.png)


<!-- Tell us how you solved the issue. -->
<!-- If there are things you want the reviewers to focus on, include them here as well. -->
<!-- This portion can be skipped if the fix is trivial. -->
@damithc @wenmogu With regards to the formatting and layout, it is dependent on the changes that have been made in PR [8836](https://github.com/TEAMMATES/teammates/pull/8836) since they use the same mechanism. [8836](https://github.com/TEAMMATES/teammates/pull/8836) needs to be merged before I can finish work on this one. The logic implementation is complete if everyone is happy with the current look.

mg14777 commented:
@wenmogu @damithc 
Instructions added and formatting completed. 
Demo: [https://teammates-mudit.appspot.com](https://teammates-mudit.appspot.com)
![image](https://user-images.githubusercontent.com/11634699/42067858-7994293c-7b40-11e8-84ba-3b8971d5229c.png)

damithc commented:
Looks good @mg14777 , based on the screenshot. I'll leave to @wenmogu to examine the actual behavior on the staging server.

mg14777 commented:
@wenmogu Added some extra tests

wenmogu commented on src/test/java/teammates/test/cases/browsertests/FeedbackRankQuestionUiTest.java:
```diff
@@ -102,107 +142,337 @@ public void testStudentSubmitAndResultsPages() throws Exception {
+                    + "At least 2 recipients have been ranked.\n"
+                    + "At most 4 recipients have been ranked.",
+                    submitPage.getRankMessage(qnNumber, 3));
 
```
what about ranking extra?

mg14777 replied:
@wenmogu sorry completely missed these comments. Can you clarify what you mean by ranking extra ?

wenmogu commented on src/main/java/teammates/common/datatransfer/questions/FeedbackRankOptionsQuestionDetails.java:
```diff
@@ -44,17 +44,7 @@ public FeedbackRankOptionsQuestionDetails() {
-        }
-
-        return instructions;
+        return null;
```
for now i will allow this to happen. but pls remove this method completely once you are done with all question types

mg14777 replied:
As far as I know we show constraint warnings only for rank and distribute points questions (please correct me if I am wrong). Just to clarify should we then remove this method from the base class `FeedbackQuestionDetails` as well ?

mg14777 commented:
@wenmogu @xpdavid  Ready for review

Updated the tests and the UI. Questions without instructions look like this:

![image](https://user-images.githubusercontent.com/11634699/43351076-c92f54d2-9204-11e8-9d36-039d8113fea7.png)

Also removed `getInstructions()` from `*QuestionDetails.java` altogether.

mg14777 commented:
@wenmogu Reminder for review

mg14777 commented:
@wenmogu A reminder. Kindly review.

mg14777 commented:
@tran-tien-dat Would you mind reviewing this as well ?

mg14777 commented:
@xpdavid Can we review and merge this ?

wenmogu commented:
Thank you @mg14777  for the good work. Looks good to me.
Soooo sorry for the late review!!!

mg14777 commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -996,42 +995,72 @@ function getMaxOptionsToBeRanked(qnNum) {
-    let areAllAnswersUnique;
     let allocatedRanks;
-    let isAllOptionsRanked;
+    let allInputsEmpty;
```
allInputs here refers to the dropdowns to allocate ranks to each recipient/option. I guess we can change it to `allDropdownsEmpty` or simply add a comment explaining the same

mg14777 commented on src/test/java/teammates/test/cases/browsertests/FeedbackRankQuestionUiTest.java:
```diff
@@ -102,114 +142,346 @@ public void testStudentSubmitAndResultsPages() throws Exception {
-        submitPage.selectResponseTextDropdown(7, 0, 2, "2");
-        submitPage.selectResponseTextDropdown(7, 0, 3, "1");
-        assertTrue("No error message expected", submitPage.getRankMessage(7, 0).isEmpty());
+        qnNumber = 1;
```
it's the same test as the one done in line 144 for all different question numbers with that property. My intention was to group these questions under a common comment

mg14777 commented on src/test/java/teammates/test/cases/browsertests/FeedbackRankQuestionUiTest.java:
```diff
@@ -102,114 +142,346 @@ public void testStudentSubmitAndResultsPages() throws Exception {
+        submitPage.selectResponseTextDropdown(qnNumber, 0, 3, "4");
+        assertFalse(submitPage.checkIfRankMessageElementExists(qnNumber, 0));
+
+        qnNumber = 7;
```
same comment as [above](https://github.com/TEAMMATES/teammates/pull/8870/files/e8bee25c946c62cffeb5a667ee9e5dc270df9092#r209447362)

mg14777 commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -1156,16 +1222,50 @@ function prepareRankQuestions() {
 
         const isRankingRecipients = $(`#rankToRecipients-${qnNum}`).val() === 'true';
 
+        // Check if feedback session is open or under preview mode
```
shouldn't it be `if (not submittable and not preview)` ?

xpdavid replied:
Opps, Yep :D

mg14777 commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.js:
```diff
@@ -1156,16 +1222,50 @@ function prepareRankQuestions() {
+        // Check if feedback session is open or under preview mode
         if (!$('#response_submit_button').is(':disabled')
             || isPreview()) {
+            const areDuplicateRanksAllowed = $(`#rankAreDuplicatesAllowed-${qnNum}`).val() === 'true';
```
umm i can't see html being generated here. If you mean line 1238 and 1239 i had this discussion with @wenmogu while working on [8836](https://github.com/TEAMMATES/teammates/pull/8836). It was suggested that all dynamic html generation should be preferably done using js since JSPs represent general structure of a web page and adding such logic to JSP would make it very specific to a particular question type.

Issue comment [here](https://github.com/TEAMMATES/teammates/pull/8836/commits/c5263a249a75471f14aa5cba54f9a99119117d86#r190470394)

![image](https://user-images.githubusercontent.com/11634699/44003683-5e61d8ba-9e74-11e8-8d10-989e84a92955.png)

mg14777 commented:
@xpdavid Ready for review

mg14777 commented:
@xpdavid Reminder for review

xpdavid commented:
Sorry for the delay. Will take a look tmr.

madanalogy commented:
Closing as the content for this PR has been migrated to Angular
